torchtitan/models/moe/moe.py

```diff
@@ -225,6 +239,70 @@ def _debug_force_load_balance_routing(
         top_scores = scores.gather(dim=1, index=selected_experts_indices)  # [N,K]
         return selected_experts_indices, top_scores
 
+    def _node_limited_routing(
+        self,
+        scores: torch.Tensor,
+        expert_bias: torch.Tensor | None,
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+        """Select top_k experts, optionally limiting to a subset of expert groups.
+
+        If num_expert_groups is set, applies node-limited routing:
+        1. Select num_limited_groups groups based on group scores
+        2. Select top_k experts only from those groups
+
+        If expert_bias is provided, it is added to scores for selection, but
+        the returned top_scores are always from the original (unbiased) scores.
+
+        Args:
+            scores: Router scores after sigmoid or softmax, shape (bs*slen, num_experts)
+            expert_bias: Optional bias for load balancing, shape (num_experts,)
+
+        Returns:
+            tuple of (selected_experts_indices, top_scores)
+                - selected_experts_indices: shape (bs*slen, top_k)
+                - top_scores: shape (bs*slen, top_k)
+        """
+        scores_for_choice = scores if expert_bias is None else scores + expert_bias
+
+        # Apply node-limited routing mask if configured
+        if self.num_expert_groups is not None:
+            if self.num_limited_groups is None:
+                raise ValueError(
+                    "num_limited_groups must be set when num_expert_groups is set"
+                )
+            if self.num_experts % self.num_expert_groups != 0:
+                raise ValueError(
+                    f"num_experts ({self.num_experts}) must be divisible by num_expert_groups ({self.num_expert_groups})"
+                )
+            experts_per_group = self.num_experts // self.num_expert_groups
+            if experts_per_group < 2:
+                raise ValueError(
+                    f"experts_per_group ({experts_per_group}) must be >= 2"
+                )
+            scores_grouped = scores_for_choice.view(
+                -1, self.num_expert_groups, experts_per_group
+            )
+            group_scores = scores_grouped.topk(2, dim=-1)[0].sum(dim=-1)
+            group_idx = torch.topk(
+                group_scores, k=self.num_limited_groups, dim=-1, sorted=False
+            )[1]
```

Comment:
```suggestion
            _, group_idx = torch.topk(
                group_scores, k=self.num_limited_groups, dim=-1, sorted=False
            )
```

For readability, it's easy to forgot why we need to slice topk()'s result

Reply:
Incorporated, thx!